fix($location): prevent infinite digest with IDN urls in Edge

Narretz opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Fix

**What is the current behavior? (You can also link to an open issue here)**
Internationalized Domain Name Urls (IDN), e.g. with an Umlaut, in Edge lead to an infinite digest on page load.

**What is the new behavior (if this is a feature change)?**
No Infdig

**Does this PR introduce a breaking change?**
No

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)

Fixes #15217

Note:
We don't test on Edge yet, but I've tested this locally with the https://github.com/nickmccurdy/karma-edge-launcher

Narretz committed:
fix($location): prevent infinite digest with IDN urls in Edge

Internationalized Domain Urls, for example Urls with Umlaut (Ä, Ö, Ü)
cause infinite digest in Edge 38.14393.0.0 because lastIndexOf doesn't
work correctly in this version. indexOf does, though.

Fixes #15217

jbedard commented:
See [my comment](https://github.com/angular/angular.js/issues/15217#issuecomment-252509972) on the bug. I'm not sure if that really matters, depends where this method is used.

jbedard commented on src/ng/location.js:
```diff
@@ -49,7 +49,7 @@ function parseAppUrl(relativeUrl, locationObj) {
 
 function startsWith(haystack, needle) {
-  return haystack.lastIndexOf(needle, 0) === 0;
+  return haystack.indexOf(needle, 0) === 0;
```
Can drop the `0` param.

petebacondarwin commented:
[core-js](https://github.com/zloirock/core-js/blob/v2.4.1/modules/es6.string.starts-with.js#L16) has an interesting patch for startsWith:

```
str.slice(index, index + search.length) === search;
```

what do you think?

gkalpak commented on src/ng/location.js:
```diff
@@ -49,7 +49,7 @@ function parseAppUrl(relativeUrl, locationObj) {
 
 function startsWith(haystack, needle) {
-  return haystack.lastIndexOf(needle, 0) === 0;
+  return haystack.indexOf(needle) === 0;
```
Maybe it is worth adding a comment to prevent someone from "optimizing" it back in the future.

gkalpak reviewed (commented):
If were concerned about perfromance, we could "sniff" if the implementation of `lastIndexOf` is broken and fall back to `indexOf` or `slice` (I am not sure which on is faster).

But this function isn't on a performance-critical path, so it might not be worth "super-optimizing".

petebacondarwin commented:
I know that micro-performance is not a good way to measure but....

Testing the following strings against `abcdefghijklmnopqrstuvwxyz` yields:

```
RegExp#test: abc x 2,574,830 ops/sec ±2.78% (76 runs sampled)
String#indexOf abc x 11,061,777 ops/sec ±2.00% (76 runs sampled)
String#slice abc x 21,384,362 ops/sec ±1.78% (78 runs sampled)
RegExp#test: xyz x 3,043,606 ops/sec ±1.78% (74 runs sampled)
String#indexOf xyz x 11,240,707 ops/sec ±1.94% (78 runs sampled)
String#slice xyz x 21,867,749 ops/sec ±1.92% (79 runs sampled)
RegExp#test: 123 x 3,072,821 ops/sec ±1.79% (79 runs sampled)
String#indexOf 123 x 12,216,397 ops/sec ±1.99% (75 runs sampled)
String#slice 123 x 22,642,025 ops/sec ±1.53% (81 runs sampled)
```

with string slice winning out easily in all cases

gkalpak commented:
> I know that micro-performance is not a good way to measure

(No buts 😛)

Narretz commented:
I can put in the slice version as startsWith, without considering any existing startsWith fn. Btw, the bug is reported here: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/9271625/ I'll add a comment to the line

Narretz committed:
fixup: faster implementation

Edge Bug url: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/9271625/

petebacondarwin commented:
LGTM
